src/Parser.js

```diff
@@ -38,28 +38,50 @@ import ParseError from "./ParseError";
  *
  * The earlier functions return ParseNodes.
  * The later functions (which are called deeper in the parse) sometimes return
- * ParseFuncOrArgument, which contain a ParseNode as well as some data about
+ * ParsedFuncOrArgOrDollar, which contain a ParseNode as well as some data about
  * whether the parsed object is a function which is missing some arguments, or a
  * standalone object which can be used as an argument to another function.
  */
 
-/** A function name or an argument to a function. */
-class ParseFuncOrArgument {
-    /**
-     * @param {ParseNode|string} result It's a string if `isFunction=true` or if
-     *     `result="$"` for switching into math mode; otherwise, it's a ParseNode
-     *     and `isFunction` must be false. If it's a function, the string should
-     *     be a name defined by defineFunction, e.g. "\\frac".
-     * @param {boolean} isFunction True when this is a function. False when it's a
-     *     function argument or "$" to switch into math mode.
-     * @param {Token} token
-     */
-    constructor(result, isFunction, token) {
-        this.result = result;
-        // Is this a function (i.e. is it something defined in functions.js)?
-        this.isFunction = isFunction;
-        this.token = token;
+/* TODO: Uncomment when porting to flow.
+type ParsedType = "fn" | "arg" | "$"
+type ParsedFunc = {|
+    type: "fn",
+    result: string, // Function name defined via defineFunction (e.g. "\\frac").
+    token?: ?Token,
+|};
+type ParsedArg = {|
+    type: "arg",
+    result: ParseNode,
+    token?: ?Token,
+|};
+type ParsedDollar = {|
+    // Math mode switch
+    type: "$",
+    result: "$",
+    token?: ?Token,
+|};
+type ParsedFuncOrArgOrDollar = ParsedFunc | ParsedArg | ParsedDollar;
+*/
+
+/**
+ * @param {ParseNode} result
+ * @param {?Token=} token
```

Comment:
Is the reason we don't pass a token sometimes is that an argument might span multiple tokens?

Reply:
Not really. Tokens can be merged from mutliple tokens. Here are the places where a token isn't supplied to the constructor:

- `arg = newArgument(this.handleUnsupportedCmd());`
- `return newArgument(new ParseNode("color", match[0], this.mode));`
- `return newArgument(new ParseNode("size", data, this.mode));`

Reply:
Actually, I can return a token in all of those existing cases.

But I also missed
- `return newArgument(new ParseNode("ordgroup", expression, this.mode, firstToken, lastToken));`
for which there isn't an obvious token text. Specifically, I don't know what to put in as the second parameter to `firstToken.range(lastToken, ???)`. I used `firstToken.text` for now and added a commit.

These changes are included in the latest commit.

Reply:
I think that's fine for now.  Eventually, we'll want to pass the token when constructing a `ParseNode` from the arg's `result`.  It may make sense for this token to include text that spans from the start of the first token to the end of the last token, but that can be in a future PR.

Reply:
I was trying to think of how to do that. While that's trivial if `firstToken` and `lastToken` come from the same lexer -- we can use `firstToken.loc.lexer.text.substring(firstToken.loc.start, lastToken.loc.end)` -- it's not so straightforward when they're not.